### PR TITLE
Make all signers operations constant time

### LINKED ISSUE #17: Improve performance and costs of revocation
Currently, `revoke` has linear complexity. It is hard to make it cheaper because the order of signers is highly relevant. It poses a danger that someone spams with a lot of signatures to increase the cost for the first signer of revoking the signature. It is quite an expensive attack but still possible.

### PR DESCRIPTION
**Pull request by jksf**
Changed `AddressSet` implementation. Previously, an array was used, and removing elements had linear complexity (preserving chronological order is crucial).
Currently, I implemented signers list as a variation of a double linked list with an index. 
This way, both adding and removing signers is constant time.
There is more hassle to iterate, but it is done in read-only calls.

I also improved tests coverage, refactored a bit and added `maxCount` parameter to `getSigners()` method. I believe we should always restrict the maximum number of elements to prevent spam slowing down the application.

I plan to improve on-chain interoperability of the contract and paging of signers, but I'll make a separate PR for these.

Closes #17

### CODE REVIEW

**Comment by jstefanski** [comment]
Closes #17

**Comment by jksf** [comment]
I attempted to make the structure single linked list, but I realized that it would make `remove()` operation linear in complexity. Having a single linked list would require traversing to remove a middle element.
Therefore, I think this is as good as we can get.

File: contracts/AddressSet.sol [hunk 1]
```diff
@@ -2,40 +2,139 @@ pragma solidity ^0.4.18;
...
+        Data storage self,
+        address element
+    )
+        internal
+        returns (bool removed)
+    {
+        require(element != address(0));
+
+        Link storage link = self.links[element];
+
```
Comment (biern) [inline contracts/AddressSet.sol]: I think this code can be made simpler by:
- checking if element exists in the beginning and returning false if not.
- Then switching the order of `head` / `tail` checking if's, like this and removing the `else if` condition:
```
        if (element == self.head) {
            self.head = link.next;
        } else {
            self.links[link.previous].next = link.next;
        }  
```
- removing the `if (removed) part`, it's safeguarded in the beginning
- and then always explicitly returning true in the end.

File: test/TestAddressSetAdd.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    function shouldAdd(address addr) private {
+        bool added = set.add(addr);
+        Assert.isTrue(added, "Should add");
+    }
+
+    function shouldNotAdd(address addr) private {
+        bool added = set.add(addr);
+        Assert.isFalse(added, "Should not add");
+    }
```
Comment (biern) [inline test/TestAddressSetAdd.sol]: I think `should` prefix is a bit misleading here it looks more like an assertion than an effect and blends with `set` assertions. Maybe:
`shouldAdd` -> `add`
`shouldNotAdd` -> `duplicateAdd`
?

File: test/TestAddressSetRemove.sol [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    function shouldRemove(address addr) private {
+        bool removed = set.remove(addr);
+        Assert.isTrue(removed, "Should remove");
+    }
+
+    function shouldNotRemove(address addr) private {
+        bool removed = set.remove(addr);
+        Assert.isFalse(removed, "Should not remove");
+    }
+}
```
Comment (biern) [inline test/TestAddressSetRemove.sol]: same as with previous `shoulds`, 
`shouldRemove` -> `remove`
`shouldNotRemove` -> `removeMissing`
?

File: test/signhash.ts [hunk 4]
```diff
@@ -115,6 +116,49 @@ contract('SignHash', accounts => {
...
+
+      const reducedCount = maxCount - 1;
+      const signers = await instance.getSigners(hash, reducedCount);
+      assert.deepEqual(signers, accounts.slice(0, reducedCount));
+    });
+
+    it('should return empty list when maxCount is zero', async () => {
+      for (const account of accounts) {
+        await instance.sign(hash, { from: account });
+      }
```
Comment (biern) [inline test/signhash.ts]: This `for` is repeated so many times it might be worth extracting it to a function

**Review by biern — COMMENTED** [review]
Impressive work! This data structure makes a lot of sense for our use cases. I also like how extensive the tests are. I left few minor comments to possible enhance the code further.

File: contracts/AddressSet.sol (hunk 1 above)
Comment (jksf) [reply contracts/AddressSet.sol]: I changed it a little differently than you suggested. You can call it micro-optimization, but I believe that low gas usage is a very important in widely used contracts. However, I tried to keep it as clean as possible.

File: test/TestAddressSetAdd.sol (hunk 2 above)
Comment (jksf) [reply test/TestAddressSetAdd.sol]: Changed to `add` and `addDuplicate`

File: test/TestAddressSetRemove.sol (hunk 3 above)
Comment (jksf) [reply test/TestAddressSetRemove.sol]: Changed to `remove` and `removeNotExisting`

File: test/signhash.ts (hunk 4 above)
Comment (jksf) [reply test/signhash.ts]: You are right. I tried to replace it with a functional style one-liner, but I could not come up with something clean and at the same time maintaining the order of accounts. I extracted the loop to a helper function. Much cleaner now.

**Comment by biern** [comment]
Yeah, I will call it micro-optimalization ;) But this code isn't meant to be read and changed a lot, but written once in an effective manner, so that's ok.

**Pull request merged**